### PR TITLE
Panzer STK: Add functionality and tests for querying the underlying PerceptMesh

### PR DESCRIPTION
**Pull request by GrahamBenHarper**
@trilinos/panzer 
@rppawlo 

## Motivation
Querying mesh family tree information from the underlying uniformly refined PerceptMesh object will allow for more advanced solving and preconditioning strategies, such as the region multigrid approach.

## Testing
It builds and 125/125 tests pass on my workstation.

## Additional Information
Ideally the `keepPerceptData` test would also check that the "Keep Percept Parent Elements" flag is working properly. Currently the lines (L358 and L382 in tPamgenMeshFactory.cpp) are commented out because I wasn't able to find a nice and simple command to do so. Perhaps @bricar could provide a suggestion?

### CODE REVIEW

File: packages/panzer/adapters-stk/test/stk_interface_test/tPamgenMeshFactory.cpp [hunk 1]
```diff
@@ -321,4 +325,63 @@ TEUCHOS_UNIT_TEST(tPamgenFactory, getMeshDimension)
...
+
+    RCP<STK_Interface> mesh = pamgenFactory->buildMesh(MPI_COMM_WORLD);
+    RCP<percept::PerceptMesh> refinedMesh = mesh->getRefinedMesh();
+
+    // check if the Percept data exists and has elements
+    TEST_ASSERT(nonnull(refinedMesh));
+    TEST_EQUALITY(refinedMesh->get_number_elements()>0,true);
+
+    // check if the parent information is stored 
+    //TEST_ASSERT(!refinedMesh->m_parent_element_field_set);
```
Comment (rppawlo) [inline packages/panzer/adapters-stk/test/stk_interface_test/tPamgenMeshFactory.cpp:382]: Can you add a check to make sure the parent element data exists? It looks like the first new option, "Keep Percept Data" is tested here but the "Keep Percept Parent Elements" option is not.

**Review by rppawlo — CHANGES_REQUESTED** [review]
@GrahamBenHarper - nice job on this! One minor comment above about testing.

**Comment by GrahamBenHarper** [comment]
```
***
*** Unit test suite ...
***


Sorting tests by group name then by the order they were added ... (time = 2.49e-06)

Running unit tests ...

0. tPamgenFactory_acceptance_UnitTest ... [Passed] (0.0147 sec)
1. tPamgenFactory_basic_UnitTest ... [Passed] (0.0165 sec)
2. tPamgenFactory_getMeshDimension_UnitTest ... [Passed] (0.002 sec)
3. tPamgenFactory_keepPerceptData_UnitTest ... 
 
 Creating pamgen mesh with parent elements.
 Test that code {pamgenFactory->setParameterList(p);} does not throw : passed
 nonnull(refinedMesh) = 1 == true : passed
 refinedMesh->get_number_elements()>0 = 1 == true : passed
 npar>0 = 0 == true : FAILED ==> /home/trilinos/jenkins/ride/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/packages/panzer/adapters-stk/test/stk_interface_test/tPamgenMeshFactory.cpp:389
 nchild>0 = 1 == true : passed
 npar>0 = 0 == true : FAILED ==> /home/trilinos/jenkins/ride/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/packages/panzer/adapters-stk/test/stk_interface_test/tPamgenMeshFactory.cpp:389
 nchild>0 = 1 == true : passed
 npar>0 = 0 == true : FAILED ==> /home/trilinos/jenkins/ride/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/packages/panzer/adapters-stk/test/stk_interface_test/tPamgenMeshFactory.cpp:389
 nchild>0 = 1 == true : passed
 
 Creating pamgen mesh without parent elements.
 Test that code {pamgenFactory->setParameterList(p);} does not throw : passed
 nonnull(refinedMesh) = 1 == true : passed
 refinedMesh->get_number_elements()>0 = 1 == true : passed
 npar = 0 == 0 = 0 : passed
 nchild>0 = 1 == true : passed
 npar = 0 == 0 = 0 : passed
 nchild>0 = 1 == true : passed
 npar = 0 == 0 = 0 : passed
 nchild>0 = 1 == true : passed
 [FAILED]  (0.0383 sec) tPamgenFactory_keepPerceptData_UnitTest
 Location: /home/trilinos/jenkins/ride/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/Trilinos/packages/panzer/adapters-stk/test/stk_interface_test/tPamgenMeshFactory.cpp:331
 

The following tests FAILED:
    3. tPamgenFactory_keepPerceptData_UnitTest ... 

Total Time: 0.0718 sec

Summary: total = 4, run = 4, passed = 3, failed = 1
```
This test failure is what I saw on my workstation as well, but the latest commit is querying the parent element information as done in percept's `Refiner.cpp`. Does anybody know why this mesh wouldn't have the parent information if I explicitly tell the breaker not to delete the parent elements?

**Comment by GrahamBenHarper** [comment]
@rppawlo @bricar any ideas why this could be happening? I can confirm that the regular refinement is working properly other than the fact that it can't find any parents on any of the topology ranks, regardless of whether `deleteParentElements()` is called or not.

**Comment by bricar** [comment]
I think this code looks correct for the purpose of this test.  It is basically what is in deleteParentElements.  There is not a function to count the parent elements, so a loop is going to be needed to find them.

Perhaps we can do a skype/teams chat this week or you can send me the steps to build and run the unit test?

**Comment by GrahamBenHarper** [comment]
@bricar I'm available any time on Tuesday and Wednesday except Tuesday from 2pm-3pm. I'd like to meet to walk through the steps briefly and see if I'm missing anything. I am also attaching my configure script that I use to run the Panzer tests. It's not necessarily a minimal configure, but it does what I need to test the Panzer build.

[configure-trilinos-panzer-tests.txt](https://github.com/trilinos/Trilinos/files/5585608/configure-trilinos-panzer-tests.txt)

Then I just run `ctest -V -R "PamgenMesh"`, and that picks the specific test that I modified.

**Comment by bricar** [comment]
@GrahamBenHarper Can you also send me the modules you loaded?  It looks like you use the sems modules but I don't think I am loading them all correctly.  Thanks, Brian.

**Comment by GrahamBenHarper** [comment]
Sorry about that @bricar! You're right; I did use the sems modules. I used the following commands:
```
module load sems-devpack-gcc/6.1.0
module load sems-ninja_fortran/1.8.2
module load sems-cmake/3.12.2
```

**Comment by GrahamBenHarper** [comment]
@rppawlo can you take a look again? I found the problem. Percept refiners delete parent elements by default, and there is a flag that has to be set in order to keep it from removing parent elements.

**Review by rppawlo — APPROVED** [review]

**Comment by rppawlo** [comment]
Thanks for adding this @GrahamBenHarper !

**Pull request merged**